### ISSUE TITLE
FlashAttention-3 and sliding window attention support

### ISSUE THREAD

**Issue by mmshad** [KempnerInstitute/KempnerForge#4]
## Current implementation

Attention lives in `kempnerforge/model/attention.py`. The `Attention` class uses PyTorch SDPA (`F.scaled_dot_product_attention`) with an `sdpa_backend` config option (`"auto"`, `"flash"`, `"efficient"`, `"cudnn"`, `"math"`); `"auto"` lets PyTorch pick.

The attention module supports:
- GQA/MQA/MHA via configurable `n_kv_heads`
- RoPE applied to Q and K before attention
- Optional QK-norm (Gemma/DeepSeek-V3 style)
- KV cache for autoregressive generation
- Document-aware causal masking for packed sequences via `doc_ids`
- Attention weight capture for interpretability (manual softmax path)

Three call paths in `Attention.forward()`:
1. `capture_attention_weights=True`: manual softmax via `_attention_with_weights`, returns attention weights (analysis only)
2. `doc_ids` provided: block-diagonal causal mask via SDPA `attn_mask`
3. Default: `is_causal=True` for training and prefill, no mask for single-token decode

## FlashAttention-3

FA3 (Dao et al., 2024) targets Hopper GPUs (H100/H200) with warp-specialization and FP8 attention. The FA3 paper reports 1.5–2.0x speedup over FA2 on H100 for bf16/fp16.

SDPA does not dispatch to FA3. It requires the standalone `flash-attn` package (Hopper build), separate from PyTorch's bundled SDPA backends.

**What needs to change:**

1. Add `flash-attn` as an optional dependency in `pyproject.toml` (guarded — Hopper-only build step).

2. In `attention.py`, add a module-level availability check:
   ```python
   try:
       from flash_attn import flash_attn_func
       _HAS_FLASH_ATTN = True
   except ImportError:
       _HAS_FLASH_ATTN = False
   ```

3. In `Attention.forward()`, add an FA3 path alongside the existing SDPA paths. `flash_attn_func` uses `(batch, seq_len, heads, head_dim)` layout (no transpose needed), so the Q/K/V reshape differs slightly on this path.

4. FA3 supports `causal=True` natively. For packed sequences with `doc_ids`, `flash_attn_varlen_func` takes `cu_seqlens` instead of an explicit mask — more efficient than constructing a full `(B, S, S)` mask.

5. Fallback: if `flash-attn` not installed, use SDPA (current behavior). Zero change for users who don't install the optional dependency.

**Integration:** `sdpa_backend` is the existing hook for attention backend selection. FA3 bypasses SDPA, so either extend `sdpa_backend` with a `"flash3"` value that routes differently, or add a sibling `attention_backend` field.

**Testing:**
- Verify FA3 output matches SDPA output within tolerance (bf16 and fp16)
- Verify FA3 with GQA produces correct output
- Verify fallback to SDPA when flash-attn not installed
- Benchmark: compare tok/s with FA3 vs SDPA on H200

## Sliding window attention

Limits each token to attend to at most W previous positions instead of the full sequence. Reduces attention memory from O(S²) to O(S·W) and attention FLOPs proportionally. Useful for long-context training (32K+ sequences) where full attention is memory-prohibitive.

Used by Mistral, Mixtral, Gemma-2, and similar architectures (some use sliding window on all layers, others alternate with full attention).

**What needs to change:**

1. Add `sliding_window: int | None = None` to `ModelConfig`. `None` means full attention (default, no behavior change).

2. Pass `sliding_window` through `TransformerBlock` to `Attention.__init__()` and store as `self.sliding_window`.

3. In `Attention.forward()`, when `sliding_window` is set:
   - For SDPA path: construct a banded causal mask (causal AND within window). PyTorch SDPA has no native sliding window parameter, so this requires an explicit `attn_mask`.
   - For FA3 path: `flash_attn_func` supports `window_size=(W, 0)` natively (left window, no right window for causal). No mask construction needed.

4. For packed sequences with `doc_ids`: the sliding window mask intersects with the document mask. Each token attends to the last W tokens within the same document only.

5. KV cache with sliding window: only cache the last W positions instead of the full sequence. Change `KVCache` to use a circular buffer when `sliding_window` is set. This reduces inference memory from O(max_seq_len) to O(W).

**Config example:**
```toml
[model]
sliding_window = 4096  # each token attends to last 4096 positions
```

**Testing:**
- Verify sliding window output matches full attention when window >= seq_len
- Verify tokens beyond window distance receive zero attention weight
- Verify sliding window works with GQA, packed sequences, and KV cache
- Verify MFU calculation accounts for reduced attention FLOPs

## MFU impact

Both changes affect `kempnerforge/metrics/mfu.py`:
- FA3: no formula change (same FLOPs, just faster hardware utilization).
- Sliding window: the `12 * L * D * S` attention term in `_dense_flops_per_token` (and `_moe_flops_per_token`) should use `min(S, W)` when sliding window is set. Without this, MFU will be under-reported for sliding window models.

**Comment by mmshad** [comment]
Split into two issues:

- #41 — FlashAttention-3 (blocked on PyTorch SDPA upstream, not adding the standalone `flash-attn` package)
- #42 — Sliding window attention (actionable, independent of FA3)

Bundling them here coupled a PyTorch-native architecture feature with an external-dep performance optimization.
